File: packages/public-api/src/SearchableFieldService.ts
import Fetch from "./Fetch.js";
import { handleResponse } from "./handleResponse.js";
import { Response, Problem, SearchableFieldResponse, CreateSearchableFieldRequest, RenameSearchableFieldRequest } from "./Types.js";

export default class SearchableFieldService {
  private readonly path = `searchable`;
  constructor(private readonly fetch: Fetch) {}

  /** Creates a Searchable field. */
  async create(request: CreateSearchableFieldRequest): Promise<Response<SearchableFieldResponse, Problem>> {
    const response = await this.fetch.post(this.path, request);
    return handleResponse(response, async () => (await response.json()) as SearchableFieldResponse);
  }

  /** Renames a Searchable. Breaking change - automatically updates the List version. */
  async rename(account_name: string, list_name: string, searchable_name: string, request: RenameSearchableFieldRequest): Promise<Response<SearchableFieldResponse, Problem>> {
    const path = `${this.path}/${encodeURIComponent(account_name)}/${encodeURIComponent(list_name)}/${encodeURIComponent(searchable_name)}`;
    const response = await this.fetch.put(path, request);
    return handleResponse(response, async () => (await response.json()) as SearchableFieldResponse);
  }

  /** Removes a Searchable. Breaking change - automatically updates the List version. */
  async remove(account_name: string, list_name: string, searchable_name: string): Promise<Response<void, Problem>> {
    const path = `${this.path}/${encodeURIComponent(account_name)}/${encodeURIComponent(list_name)}/${encodeURIComponent(searchable_name)}`;
    const response = await this.fetch.delete(path);
    return handleResponse(response, () => undefined);
  }
}

File: dotnet/SearchStack.PublicApi/Services/SearchableFieldService.cs
using System.Threading;
using System.Threading.Tasks;
using static SearchStack.PublicApi.Url;

namespace SearchStack.PublicApi;

/// <summary>Searchable field create / rename / remove operations.</summary>
public sealed class SearchableFieldService
{
    private const string Path = "searchable";
    private readonly Transport _transport;

    internal SearchableFieldService(Transport transport) => _transport = transport;

    /// <summary>Creates a Searchable field.</summary>
    public Task<Response<SearchableFieldResponse>> CreateAsync(CreateSearchableFieldRequest request, CancellationToken cancellationToken = default) =>
        _transport.PostAsync<SearchableFieldResponse>(Path, request, cancellationToken);

    /// <summary>Renames a Searchable. Breaking change - automatically updates the List version.</summary>
    public Task<Response<SearchableFieldResponse>> RenameAsync(string accountName, string listName, string searchableName, RenameSearchableFieldRequest request, CancellationToken cancellationToken = default) =>
        _transport.PutAsync<SearchableFieldResponse>($"{Path}/{Enc(accountName)}/{Enc(listName)}/{Enc(searchableName)}", request, cancellationToken);

    /// <summary>Removes a Searchable. Breaking change - automatically updates the List version.</summary>
    public Task<Response> RemoveAsync(string accountName, string listName, string searchableName, CancellationToken cancellationToken = default) =>
        _transport.DeleteAsync($"{Path}/{Enc(accountName)}/{Enc(listName)}/{Enc(searchableName)}", cancellationToken);
}
